src/traits.rs

```diff
@@ -699,6 +699,17 @@ pub trait RemMixed<Reductor>: Sized {
     fn rem_mixed(&self, reductor: &NonZero<Reductor>) -> Reductor;
 }
 
+/// Modular reduction from a larger value `T`.
+///
+/// This can be seen as fixed modular reduction, where the modulus is fixed at compile time
+/// by `Self`.
+///
+/// For modular reduction with a variable modulus, use [`Rem`].
+pub trait Reduce<T>: Sized {
+    /// Reduces `self` modulo `Modulus`.
+    fn reduce(value: T) -> Self;
```

Comment:
I'd suggest a borrow here. It's a transformation of the input, not something that takes ownership. It's also potentially a secret, so enforcing a borrow leaves the original value zeroizable.

See also: https://rust-lang.github.io/api-guidelines/flexibility.html?highlight=clone#caller-decides-where-to-copy-and-place-data-c-caller-control

```suggestion
    fn reduce(value: &T) -> Self;
```

Reply:
Actually, here are some examples in `RustCrypto` that take things by value in reduce operations:

- https://docs.rs/p256/0.14.0-pre.9/p256/struct.Scalar.html#method.reduce
- https://docs.rs/p384/0.14.0-pre.9/p384/struct.Scalar.html#method.reduce

If we keep the input by value we can do both these implementations and implementations on referenced byte slices or arrays.

Reply:
It seems to me like both P256 and P384 would be OK with being passed a reference instead.

Their current sources [[P256](https://docs.rs/p256/0.14.0-pre.9/src/p256/arithmetic/scalar.rs.html#626-630)] 
```rust
impl Reduce<U256> for Scalar {
    fn reduce(w: U256) -> Self {
        let (r, underflow) = w.borrowing_sub(&NistP256::ORDER, Limb::ZERO);
        let underflow = Choice::from((underflow.0 >> (Limb::BITS - 1)) as u8);
        Self(U256::conditional_select(&w, &r, !underflow))
    }
}
```
and [[P384](https://docs.rs/p384/0.14.0-pre.9/src/p384/arithmetic/scalar.rs.html#233-237)]
```rust
impl Reduce<U384> for Scalar {
    fn reduce(w: U384) -> Self {
        let (r, underflow) = w.borrowing_sub(&NistP384::ORDER, Limb::ZERO);
        let underflow = Choice::from((underflow.0 >> (Limb::BITS - 1)) as u8);
        Self::from_uint_unchecked(U384::conditional_select(&w, &r, !underflow))
    }
}
```

Reply:
Indeed. The claim is that having the function signature by value allows both `impl Reduce<U384>` and `impl Reduce<&U384>`, whereas the proposed change invalidates `impl Reduce<U384>` without adding anything.